allennlp/data/iterators/multiprocess_iterator.py

```diff
@@ -0,0 +1,152 @@
+from typing import Iterable, Iterator, List, Optional
+import logging
+
+from torch.multiprocessing import Process, Queue, get_logger
+
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.util import lazy_groups_of
+from allennlp.data.instance import Instance
+from allennlp.data.iterators.data_iterator import DataIterator, TensorDict
+from allennlp.data.dataset import Batch
+from allennlp.data.vocabulary import Vocabulary
+
+logger = get_logger()  # pylint: disable=invalid-name
+logger.setLevel(logging.INFO)
+
+def _create_tensor_dicts(input_queue: Queue,
+                         output_queue: Queue,
+                         max_instances_in_memory: int,
+                         batch_size: int,
+                         vocab: Vocabulary) -> None:
+    """
+    Pulls at most ``max_instances_in_memory`` from the input_queue,
+    groups them into batches of size ``batch_size``, converts them
+    to ``TensorDict`` s, and puts them on the ``output_queue``.
+    """
+    instances: List[Instance] = []
+
+    def make_batches() -> None:
+        for batch_instances in lazy_groups_of(iter(instances), batch_size):
+            batch = Batch(batch_instances)
+
+            if vocab is not None:
+                batch.index_instances(vocab)
+
+            padding_lengths = batch.get_padding_lengths()
+            tensor_dict = batch.as_tensor_dict(padding_lengths)
+
+            output_queue.put(tensor_dict)
+        instances.clear()
+
+    while True:
+        instance = input_queue.get()
+        if instance is None:
+            # This means there are no more instances, so
+            # create one last batch (if necessary),
+            # put None on the queue, and then break out of the loop.
+            if instances:
+                make_batches()
+            output_queue.put(None)
+            break
+        else:
+            instances.append(instance)
+            if len(instances) >= max_instances_in_memory:
+                make_batches()
+
+def _queuer(instances: Iterable[Instance],
```

Comment:
Is there a way to remove the `_queuer` if we are also using the `MultiprocessDatasetReader ` by passing the reader queue directly to the `MultiprocessIterator`?  E.g. the `MultiprocessIterator` can iterate over either an iterable of instances or a queue filled with instances?